drivers: flash: shell: add "flash copy" command

cfriedt opened the pull request:
Add a `flash copy` command, capable of copying a region in one flash device to a region on the same or another flash device. The caller is responsible for erasing the destination prior to copying, if that is necessary.

This is useful for evaluating mcuboot on devices with little on-chip resources that are not capable of running more elaborate image management services (e.g. via bluetooth or networking).

Additionally, it's useful for evaluating mcuboot on devices with one or more images stored on external spi flash.

The `flash copy` command syntax is
```shell
flash copy <src_dev> <dst_dev> <src_offs> <dest_offs> <size>
```

E.g.
```shell
flash copy flash@0 flash-controller@abcd1234 0x1234 0x5678 21012
Copied 21012 bytes from flash@0:1234 to flash-controller@abcd1234:5678
```

cfriedt commented:
Probably should check to verify that the regions do not overlap before doing the write (which would be erm... tricky), and also should check the CRCs of source and destination before doing the write (to see if they are equivalent).

Are we missing a generic `flash_copy()` algorithm in Zephyr? This could really be 1 function call shared by lots of different code.

Commit:
drivers: flash: provide a generic flash_copy() algorithm

Provide a generic flash_copy() algorithm that is capable of
copying from one flash device to another or within different
regions of the same flash device.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
drivers: flash: shell: add "flash copy" command

Add a flash copy command, capable of copying a region in one
flash device to a region on the same or another flash device. The
destination is erased prior to copying.

This is useful for evaluating mcuboot on devices with little
on-chip resources, or devices that are incapable of running more
elaborate image management services (e.g. via bluetooth or
networking).

Additionally, it's useful for evaluating mcuboot on devices with
one or more images stored on external spi flash.

The command syntax is
flash copy <src_dev> <dst_dev> <src_offs> <dest_offs> <size>

E.g.
flash copy flash@0 flash-controller@abcd1234 0x1234 0x5678 21012
Copied 21012 bytes from flash@0:1234 to \
  flash-controller@abcd1234:5678

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: drivers: flash: add flash_copy() test

Add a test for coverage of flash_copy()

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

nordicjm commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+	off_t end;
+	uint32_t _buf[4];
+
+	if (!IS_ENABLED(CONFIG_CRC)) {
```
would need an #ifdef here otherwise calling `crc32_ieee_update` will give undefined function

cfriedt replied:
I'll double check that, but I don't believe that to be the case.

cfriedt replied:
Fixed (crc is removed)

nordicjm commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+	       off_t dst_offset, off_t size, bool verify, uint32_t *buf, size_t buf_size)
+{
+	int ret;
+	uint32_t _buf[4];
```
description says 16 byte buffer

cfriedt replied:
4 * 4 = 16

nordicjm commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+		     uint32_t *buf, size_t buf_size)
+{
+	off_t end;
+	uint32_t _buf[4];
```
description says 16 byte buffer

cfriedt replied:
4 * 4 = 16

cfriedt replied:
Fixed - buffer is external now.

cfriedt commented:
It's weird how GH duplicated my responses as comments. Oh well

de-nordic commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+
+static bool off_add_overflow(off_t offset, off_t size, off_t *result)
+{
+	if (sizeof(off_t) == sizeof(uint32_t)) {
+		uint32_t end;
+
+		/* account for signedness of off_t due to lack of s32_add_overflow() */
+		if (u32_add_overflow((uint32_t)offset, (uint32_t)size, &end) ||
+		    (end > (UINT32_MAX >> 1))) {
+			return true;
+		}
+	} else if (sizeof(off_t) == sizeof(uint64_t)) {
+		uint64_t end;
+
+		/* account for signedness of off_t due to lack of s64_add_overflow() */
+		if (u64_add_overflow((uint64_t)offset, (uint64_t)size, &end) ||
+		    (end > (UINT64_MAX >> 1))) {
+			return true;
+		}
+	}
```
This is overly complicated.
Since `off_t` is signed, then 
```
#define OFF_T_MIN ((off_t)1 << (sizeof(off_t) * 8  -1))
#define OFF_T_MIN (~(OFF_T_MIN))
/* At this point both offset and  size should be >= 0, as they should be verified by caller */
if (OFF_T_MAX - offset < size) {
    return true;
}

*result = offset + size;
return false;
```

then there is the question whether the function is even needed as this can just be inlined.
This is the same approach as taken in Integer Security check from "Secure Coding in C and C++" or chapter INT32-C from "The CERT C Secure Coding Standard".
There should be check for offset >= 0 and size >=0 but this is assumed.

cfriedt replied:
I haven't yet addressed this, but will take a look again tomorrow most likely.

de-nordic commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+
+	BUILD_ASSERT((sizeof(off_t) == sizeof(uint32_t)) || (sizeof(off_t) == sizeof(uint64_t)));
+
+	if ((a_start < 0) || (a_size < 0) || (b_start < 0) || (b_size < 0)) {
+		return -EINVAL;
+	}
```
That is why we need RFC for unsigned only `off_t` for storage, at least at the device level. I do not know any device that allows negative offsets to address bytes. Ok, I know one: that is sometimes in SPI NOR device, but it is rather side effect of overlap when crossing device size.

cfriedt replied:
> That is why we need RFC for unsigned only `off_t` for storage, at least at the device level. 

At the device level yes, but certain API calls (e.g. `lseek()`) do require the ability to seek backward, so a signed `off_t` is still necessary.

de-nordic replied:
Actually you do not if you have other parameter to indicate direction, something that `lseek` does not do

cfriedt replied:
`lseek()` doesn't have a direction parameter. It accepts `SEEK_SET`,  `SEEK_CUR`, and `SEEK_END` (for beginning, current position, and end) and then the `off_t` is relative to those locations.

de-nordic replied:
> `lseek()` doesn't have a direction parameter. It accepts `SEEK_SET`, `SEEK_CUR`, and `SEEK_END` (for beginning, current position, and end) and then the `off_t` is relative to those locations.

Yes, I have noticed
![image](https://github.com/user-attachments/assets/64f20f24-eb53-4e8f-b9ab-bf56990b983b)

de-nordic commented on include/zephyr/drivers/flash.h:
```diff
@@ -583,6 +583,84 @@ static inline const struct flash_parameters *z_impl_flash_get_parameters(const s
+ * @retval -EINVAL if any value provided is negative.
+ * @retval -EOVERFLOW if either region overflows the address space.
+ */
+int flash_regions_overlap(off_t a_start, off_t a_size, off_t b_start, off_t b_size);
```
Can we use "range" instead of "region"? It is just that region somehow reminds me of something datasheet may describe as having specific properties in a device, like write-once region, protected region.

cfriedt replied:
Sure

de-nordic commented on include/zephyr/drivers/flash.h:
```diff
@@ -583,6 +583,84 @@ static inline const struct flash_parameters *z_impl_flash_get_parameters(const s
+ *
+ * @see crc32_ieee_update
+ */
+int flash_region_crc(const struct device *dev, off_t offset, off_t size, uint32_t *crc,
```
I am not convinced regarding passing the user buffer.
What would be the use case? Re-use of some buffer user already has?

cfriedt replied:
removed

de-nordic commented on include/zephyr/drivers/flash.h:
```diff
@@ -583,6 +583,84 @@ static inline const struct flash_parameters *z_impl_flash_get_parameters(const s
+ * @retval 0 if the regions do not overlap.
+ * @retval 1 if the regions overlap.
+ * @retval -EINVAL if any value provided is negative.
+ * @retval -EOVERFLOW if either region overflows the address space.
```
I think this is something we should take under RFC. I think that all the other functions, the Flash API calls from drivers, return `-EINVAL` in this case. Is this something we should change then?

cfriedt replied:
I can change it to `-EINVAL`

de-nordic commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+		}
+	}
+
+	/* This fails in testing for qemu_x86 iirc, with a flash simulator */
```
Eh, I will need to investigate that. This should not happen.

cfriedt commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+	}
+
+	/* This fails in testing for qemu_x86 iirc, with a flash simulator */
+	(void)flash_erase(dst_dev, dst_offset, size);
```
The original implementation of this was using copy by erase size.

For some reason, I couldn't get the erase page size easily out of DT but I could revisit.

I think it would be fine to switch back to that.

cfriedt replied:
Fixed! (caller must erase prior to copy)

cfriedt commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+		buf_size = sizeof(_buf);
+	}
+
+	if ((buf_size & (sizeof(uint32_t) - 1)) != 0) {
+		LOG_DBG("buffer size is not a multiple of uint32_t size");
+		return -EINVAL;
+	}
```
I think this was more to do with the write size for the devices I was using. One had a minimum write size of 4 bytes and the other, 8.

cfriedt commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+		 "%s buf of size %u is too small for write_size of %u", __func__,
+		 (uint32_t)sizeof(buf), write_size);
+
+	if (verify && IS_ENABLED(CONFIG_CRC)) {
+		ret = flash_region_crc(src_dev, src_offset, size, &crc_src, buf, buf_size);
+		if (ret < 0) {
+			LOG_DBG("%s failed: %d", "flash_region_crc()", ret);
+			return ret;
+		}
+
+		ret = flash_region_crc(dst_dev, dst_offset, size, &crc_dst, buf, buf_size);
+		if (ret < 0) {
+			LOG_DBG("%s failed: %d", "flash_region_crc()", ret);
+			return ret;
+		}
+
+		if (crc_dst == crc_src) {
+			/* source and destination regions already have the same contents */
+			return 0;
+		}
+	}
+
```
So leave verification external? I can do that, sure.

Originally, my algorithm (prior to this PR) calculated the CRC one erase block at a time (and so would bail early if a write error was detected). I couldn't find a good way to get the erase block size in general though, and as you said, some parts don't even have that function, so I changed things prior to my original submission.

I'll remove the CRC part.

de-nordic replied:
You could still do that with another utility combo-function that could take the validation function as a parameter.
This way you can still do the `flash_copy` as a syscall, without executing user provided function in syscall space, but at the same time provide means for user to have any validation function they want.

cfriedt replied:
Fixed! (removed crc function)

cfriedt commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+	       off_t dst_offset, off_t size, bool verify, uint32_t *buf, size_t buf_size)
+{
+	int ret;
+	uint32_t _buf[4];
```
I wonder if it's possible for it to be done in Devicetree macros.

Something compile-time constant would be ideal. Of course it would only apply to the devices known at build time.

de-nordic replied:
The problem is that you can have drivers that do not even touch DTS, so you have to have an option to  override that in Kconfig too.
I also do not like picking such things from DTS, because it basically means that the value leaks through several layers, and ends up having unforeseen consequences.

cfriedt replied:
Fixed (caller should erase prior to copy)

cfriedt commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+			LOG_DBG("crc mismatch: expected: %08x, actual: %08x", crc_src, crc_dst);
+			return -EIO;
+		}
+	}
```
Fixed!

de-nordic commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,111 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+			memset(buf, 0, write_size);
+		}
+		bytes_read = MIN(MAX(bytes_left, write_size), buf_size);
+		ret = flash_read(src_dev, src_offset + offs, buf, bytes_read);
```
Don't you want to make the `flash_copy` a syscall? This `flash_read` will got through the syscall gate every loop iteration, in case when `CONFIG_USERSPACE=y`;
or you are not making it syscall to prevent spending too much time in syscall while doing the copy?

cfriedt replied:
Sure - I think there are a couple of changes left to do here still, so it's not done by any means.

cfriedt replied:
Fixed!

cfriedt commented:
* included `<zephyr/internal/syscall_handler.h>` for userspace support in `z_vrfy_flash_copy()`

cfriedt commented:
* used non-const array, which somehow fixed `qemu_cortex_m0`

cfriedt commented:
* cleaned-up commit messages

cfriedt commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+				(long)(src_offset + offs), ret);
+			return ret;
+		}
+	}
```
Fixed!

cfriedt commented:
* fixed formatting issues

nordicjm commented on tests/drivers/flash/common/src/main.c:
```diff
@@ -253,5 +253,94 @@ ZTEST(flash_driver, test_flash_flatten)
+	 *  - buf is NULL
+	 *  - buf size is invalid
+	 */
+	struct arg_exp {
```
you would need to make this packed for it to work on m0

cfriedt replied:
packed seems kind of odd in this context. I think the issue is that the stack is too small.

nordicjm replied:
Seems it might be stack after all

cfriedt replied:
Instead of creating a large array of data on the stack, I simply created an `_inner()` function, and passed arguments directly as literals, so that data should just be in .text now.

Commit:
tests: flash: common: randomize before each test

test_flash_erase() requires that the expected[] array contains
pseudo-random data. However, the expected[] array would only
be initialized once before all tests are run using the setup
callback.

Instead, use the before() callback to randomize data before
each test in the suite, since there is otherwise no guarantee
that test_flash_erase() will be run directly after the
expected[] array has been randomized.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
@de-nordic, @nordicjm - please take a look

nordicjm commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,112 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+	     offs += bytes_read, bytes_left -= bytes_read) {
+
+		if (bytes_left < write_size) {
+			memset(buf, 0, write_size);
```
this should probably be setting it to erase value of flash device

cfriedt commented:
@nordicjm - had to wrap one long line. Please re-ack when you have a moment.

d3zd3z commented on drivers/flash/flash_util.c:
```diff
@@ -80,3 +87,116 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+}
+
+/* note: caller must first check for overflow */
+static inline bool flash_ranges_overlap(off_t a_start, off_t a_size, off_t b_start, off_t b_size)
```
We need to check more than just that the ranges don't overlap, but that the entirety of source and dest are in different erase regions. Otherwise, you're likely to just end up with a bunch of flash full of the erase value.

cfriedt replied:
If I'm interpreting this correctly, I think this basically involves rounding the star addresses down by a write block size and rounding end addresses up by the same.

Note: the caller is responsible for erasing prior to copying, if necessary.

cfriedt replied:
@d3zd3z - my intuition says that the start start address of the destination should be a multiple of the write block size, otherwise it would be an error, and then rounding up the end address to a write block size (padded with erase values) should be sufficient. I think, as long as the range does not overlap (with a possibly rounded end address), that should work. I could potentially add a test for each of those conditions as well.

Does that make sense?

cfriedt commented:
> It isn't clear to me where the erase is happening

Erase (if necessary) is called by the user separately. Both that and write verification were originally included here but were later externalized based on reviews.
